### PR TITLE
datapath: bigtcp: Fix the IPv4 BIG TCP may not work

### PR DESCRIPTION
**Pull request by haiyuewa**
The kernel will also update the IPv4 GRO/GSO setting if the new value of
"gso/gro_max_size" isn't greater than 65536:
  1. ip link set dev ens787np0 gso_max_size 130000
  2. ip link set dev ens787np0 gso_ipv4_max_size 128000
  3. ip -d -j link show dev ens787np0 | jq -c '.[0].gso_ipv4_max_size'
     128000
  4. ip link set dev ens787np0 gso_max_size 65536
  5. ip -d -j link show dev ens787np0 | jq -c '.[0].gso_ipv4_max_size'
     65536

Cilium will call IPv4 GRO/GSO setup firstly, then call IPv6. To handle the
above issue, setup the IPv6 GRO/GSO firstly, then IPv4 GRO/GSO.

### CODE REVIEW

File: pkg/datapath/linux/bigtcp/bigtcp.go [hunk 1]
```diff
@@ -219,31 +219,31 @@ func InitBIGTCP(bigTCPConfig *Configuration) {
 		// modification and end up with an error, so best to be conservative
 		// and always reset all on error
 		modifiedDevices = append(modifiedDevices, device)
-		if haveIPv4 {
-			err = setGROGSOIPv4MaxSize(device,
-				bigTCPConfig.groIPv4MaxSize, bigTCPConfig.gsoIPv4MaxSize)
+		if haveIPv6 {
```
Comment (PlatformLC) [inline pkg/datapath/linux/bigtcp/bigtcp.go:225]: nit: should we leave an comment for this tricky order directly in code? Or hide such logic in setGROGSOIPv6MaxSize to make sure IPv6 operation will not impact IPv4 values.

File: pkg/datapath/linux/bigtcp/bigtcp.go [hunk 2]
```diff
@@ -219,31 +219,31 @@ func InitBIGTCP(bigTCPConfig *Configuration) {
...
 				option.Config.EnableIPv4BIGTCP = false
 				option.Config.EnableIPv6BIGTCP = false
 				break
 			}
-			log.WithField("device", device).Infof("Setting IPv6 gso_max_size to %d and gro_max_size to %d",
-				bigTCPConfig.gsoIPv6MaxSize, bigTCPConfig.groIPv6MaxSize)
+			log.WithField("device", device).Infof("Setting IPv4 gso_max_size to %d and gro_max_size to %d",
+				bigTCPConfig.gsoIPv4MaxSize, bigTCPConfig.groIPv4MaxSize)
 		}
 	}
```
Comment (PlatformLC) [inline pkg/datapath/linux/bigtcp/bigtcp.go:252]: nit: should we also modify the IPv4/IPv6 order for below err handling?

File: pkg/datapath/linux/bigtcp/bigtcp.go (hunk 1 above)
Comment (haiyuewa) [reply pkg/datapath/linux/bigtcp/bigtcp.go:225]: This will make 'setGROGSOIPv6MaxSize' not so clean, by my understanding. And this logic is in kernel, git commit message should be fine, no bugs, then peope will be fine about this. ; - )

File: pkg/datapath/linux/bigtcp/bigtcp.go (hunk 2 above)
Comment (haiyuewa) [reply pkg/datapath/linux/bigtcp/bigtcp.go:252]: 64KB will be OK. Just less code change to make things work.

File: pkg/datapath/linux/bigtcp/bigtcp.go (hunk 1 above)
Comment (haiyuewa) [reply pkg/datapath/linux/bigtcp/bigtcp.go:225]: After thinking more, looks like make sense to add some comments. Thanks!

File: test/bigtcp/test.sh [hunk 3]
```diff
@@ -43,7 +43,7 @@ helm install cilium ${HELM_CHART_DIR} \
 kubectl -n kube-system rollout status ds/cilium --timeout=5m
 
 # check if BIG TCP is initialized
-kubectl -n kube-system logs ds/cilium 2>&1 | grep "Setting up IPv6 BIG TCP"
+kubectl -n kube-system logs ds/cilium 2>&1 | grep "Setting up BIG TCP"
```
Comment (borkmann) [inline test/bigtcp/test.sh]: thanks for the fix. this could probably be submitted as a separate PR and we can already merge it.

File: pkg/datapath/linux/bigtcp/bigtcp.go (hunk 1 above)
Comment (borkmann) [reply pkg/datapath/linux/bigtcp/bigtcp.go:225]: Thanks for the fix, @haiyuewa! Were you able to test it on ice with localy tweaking default{GSO,GRO}MaxSize? Would be curious what numbers you see before/after on ice.

Do you plan to follow-up with a knob for tweaking internal default{GSO,GRO}MaxSize?

File: test/bigtcp/test.sh (hunk 3 above)
Comment (haiyuewa) [reply test/bigtcp/test.sh]: Done: https://github.com/cilium/cilium/pull/26377

File: pkg/datapath/linux/bigtcp/bigtcp.go (hunk 1 above)
Comment (haiyuewa) [reply pkg/datapath/linux/bigtcp/bigtcp.go:225]: I will re-submit with new logs after tweaking the MAX size.

And yes, if you don't start to code it. ;-) It should be about tweaking internal bigTCP{GRO/GSO}MaxSize, right ?

Comment (borkmann) [reply pkg/datapath/linux/bigtcp/bigtcp.go:225]: Back in IPv6 testing, @NikAleksandrov found the 196608 to be a sweet spot for mlx5.

I think for ice, we could read out link.Attrs().TSOMaxSize and probably determine this automatically even without knob.

For example, if the default{GSO,GRO}MaxSize fails to be set, then if link.Attrs().TSOMaxSize is still larger than 65536 we know that the device supports BIG TCP.

So if larger than 65536, we could take min(default{GSO,GRO}MaxSize, link.Attrs().TSOMaxSize) to figure it out.

Comment (borkmann) [reply pkg/datapath/linux/bigtcp/bigtcp.go:225]: @haiyuewa if you could help with testing on ice, I could code this up. But lets get first your fix from here in, so that we don't run into merge conflict.

Comment (haiyuewa) [reply pkg/datapath/linux/bigtcp/bigtcp.go:225]: @borkmann Looks like more things need to be considered. Sure, I can help on test it.

Comment (borkmann) [reply pkg/datapath/linux/bigtcp/bigtcp.go:225]: > Looks like more things need to be considered.

Could you elaborate?

**Review by borkmann — APPROVED** [review]
Awesome, looks good to me! Thanks!

File: pkg/datapath/linux/bigtcp/bigtcp.go (hunk 1 above)
Comment (haiyuewa) [reply pkg/datapath/linux/bigtcp/bigtcp.go:225]: > > Looks like more things need to be considered.
> 
> Could you elaborate?

I just thought it needs to do more things about "Back in IPv6 testing, @NikAleksandrov found the 196608 to be a sweet spot for mlx5." It's my misunderstanding.

Please let me know if the code is ready for ice test. Thanks.

Comment (borkmann) [reply pkg/datapath/linux/bigtcp/bigtcp.go:225]: Btw, I just checked the veth side, and here we should be good as v6 is adjusted before v4: https://github.com/cilium/cilium/pull/26172/commits/0ab201b39b55ab36e6bd6ba98d60f5da851afe73

**Review by ldelossa — APPROVED** [review]
